src/pretty.rs

```diff
@@ -0,0 +1,471 @@
+use crate::term::{BinaryOp, MetaValue, RecordAttrs, RichTerm, Term};
+use crate::types::{AbsType, Types};
+pub use pretty::{DocAllocator, DocBuilder, Pretty};
+
+impl<'a, A: Clone + 'a> NickelAllocatorExt<'a, A> for pretty::BoxAllocator {}
+
+trait NickelAllocatorExt<'a, A: 'a>: DocAllocator<'a, A> + Sized
+where
+    Self::Doc: Clone,
+    A: Clone,
+{
+    fn metadata(self: &'a Self, mv: &MetaValue, with_doc: bool) -> DocBuilder<'a, Self, A> {
+        if let Some(types) = &mv.types {
+            self.text(": ")
+                .append(types.types.clone().pretty(self))
+                .append(self.line())
+        } else {
+            self.nil()
+        }
+        .append(
+            self.intersperse(
+                mv.contracts
+                    .iter()
+                    .map(|c| self.text("| ").append(c.to_owned().types.pretty(self))),
+                self.line().clone(),
+            ),
+        )
+        .append(if mv.priority == crate::term::MergePriority::Default {
+            self.line().append(self.text("| default"))
+        } else {
+            // if last field is not default, it will add a optional line break before the `=` sign
+            self.nil()
+        })
+        .nest(2)
+        .group()
+    }
+
+    fn atom(&'a self, rt: &RichTerm) -> DocBuilder<'a, Self, A> {
+        if rt.as_ref().is_atom() {
+            rt.to_owned().pretty(self)
+        } else {
+            rt.to_owned().pretty(self).parens()
+        }
+    }
+}
+
+impl<'a, D, A> Pretty<'a, D, A> for BinaryOp
+where
+    D: NickelAllocatorExt<'a, A>,
+    D::Doc: Clone,
+    A: Clone + 'a,
+{
+    fn pretty(self, allocator: &'a D) -> DocBuilder<'a, D, A> {
+        use BinaryOp::*;
+        match self {
+            Plus() => allocator.text("+"),
+            Sub() => allocator.text("-"),
+
+            Mult() => allocator.text("*"),
+            Div() => allocator.text("/"),
+            Modulo() => allocator.text("%"),
+
+            Eq() => allocator.text("=="),
+            LessThan() => allocator.text("<"),
+            GreaterThan() => allocator.text(">"),
+            GreaterOrEq() => allocator.text(">="),
+            LessOrEq() => allocator.text("<="),
+
+            Merge() => allocator.text("&"),
+
+            op => allocator.as_string(format!("%{:?}%", op).to_lowercase()),
+        }
+    }
+}
+
+impl<'a, D, A> Pretty<'a, D, A> for RichTerm
+where
+    D: NickelAllocatorExt<'a, A>,
+    D::Doc: Clone,
+    A: Clone + 'a,
+{
+    fn pretty(self, allocator: &'a D) -> DocBuilder<'a, D, A> {
+        use Term::*;
+
+        match self.as_ref() {
+            Null => allocator.text("null"),
+            Bool(v) => allocator.as_string(v),
+            Num(v) => allocator.as_string(v),
+            Str(v) => allocator.as_string(v),
+            StrChunks(chunks) => allocator
+                .text("m%\"")
+                .append(allocator.intersperse(
+                    chunks.iter().map(|c| {
```

Comment:
You need to reverse chunks, as they are stored in the the reverse direction (as a stack of chunks to be popped). For example, the current implementation gives:

```console
$ nickel expend <<< ' "unkown flag %{flag}" '
m%"%{flag}
unkown flag "%m
```

Reply:
Also, you may want to special case when `chunks` is of size 1 to use standard string delimiters, otherwise `m%"` is always used, even for small strings.